### DISCUSSION TITLE
Passport Protocol Debug

### DISCUSSION THREAD

**Discussion by diego-rapoport** [TypedProject/ts-express-decorators#1912] [Q&A]
So, I'm unable to debug my protocol with `console.log` neither with `$log...`. Writing on `$onInstall`or `$onVerify`.
Do I need to configure something on the `Server.ts`?
```
@Protocol<IStrategyOptions>({
  name: 'login',
  useStrategy: BasicStrategy,
  settings: {
    usernameField: 'email',
    passwordField: 'senha',
  }
})
export class LoginProtocol implements OnVerify, OnInstall {
  @Inject()
  private usuarioServico: UsuarioService;

  $onInstall(strategia: Strategy): void {
    console.log('STRATEGIA = ', strategia)
  }

  async $onVerify(@Req() requisicao: Req, @Arg(0) email: string, @Arg(1) senha: string) {
     $log.warn('REQ = ', requisicao)
    const conferido = await this.usuarioServico.confereSenha({ email: email, senha: senha })
    if (!conferido) return false
    return true
  }
}
```

**Comment by Romakita** [comment]
Have you imported the protocol with an es import in server.ts?

import "./protocols/MyProtocol"

**Comment by diego-rapoport** [comment]
Yes I did
`import "@protocols/LoginProtocol";`

**Reply by Romakita** [reply]
Add a console before the class to be sure or onnthe constructor. After the $onInstall is a recent method and the $onVerify is called when the strategy is used on an endpoint. Also check your tsed dependencies. All modules must have the same version

**Reply by diego-rapoport** [reply]
Ok so I can see the log before the class and now I saw where my log of strategy went. So this is working properly.

**Reply by diego-rapoport** [reply]
My tsed dependencies:
```
  "dependencies": {
    "@tsed/ajv": "^6.102.8",
    "@tsed/common": "^6.102.8",
    "@tsed/core": "^6.102.8",
    "@tsed/di": "^6.102.8",
    "@tsed/engines": "^1.1.5",
    "@tsed/exceptions": "^6.102.8",
    "@tsed/json-mapper": "^6.102.8",
    "@tsed/logger": "^6.1.0",
    "@tsed/logger-file": "^6.1.0",
    "@tsed/mongoose": "^6.102.8",
    "@tsed/passport": "^6.114.13",
    "@tsed/platform-express": "^6.102.8",
    "@tsed/schema": "^6.102.8",
```

**Reply by diego-rapoport** [reply]
> Ok so I can see the log before the class and now I saw where my log of strategy went. So this is working properly.

Still can't debug inside `$onVerify`

**Reply by Romakita** [reply]
The $onVerify is called only when the strategy is used on an endpoint.

Here an example:

```typescript
@Controller("/auth")
@Scope(ProviderScope.SINGLETON)
export class AuthCtrl {
  @Post("/login")
  @Authenticate("login")
  login(@Req() req: Req, @BodyParams("email") email: string, @BodyParams("password") password: string) {
    // FACADE
    return req.user;
  }
}
```
A simple postman is enough to trigger the onVerify method.
See you
Romain

**Reply by diego-rapoport** [reply]
Sorry, I must be understanding and doing something very wrong because that's exactly what I did. I just added a `console.log` where you wrote the `// FACADE` and I'm still getting nothing. Also I just get `Unauthorized` as response, even giving right credentials to my backend.

I also couldn't understand very well the difference between `Strategy` and `BasicStrategy`. Strategy gives me 401 and Basic gives me 400

**Reply by Romakita** [reply]
Basic strategy, use header to authenticate account while login strategy use a login webservice.

if you have a 401, it means passeport handle the request but you haven’t sent the correct information to execute the strategy.

Please look these unit test: https://github.com/tsedio/tsed-example-passportjs/blob/master/test/integration/controllers/auth/auth.spec.ts

**Reply by Romakita** [reply]
I just updated the example with the latest version. All unit test works:
[https://github.com/tsedio/tsed-example-passportjs](https://github.com/tsedio/tsed-example-passportjs)

**Reply by diego-rapoport** [reply]
Thank you very much for the reply. Sure will work on that looking into these examples.
